File: examples/eilmer/2D/oblique-detonation-wave/odw.lua
-- odw.lua
-- Powers and Aslam oblique detonation wave verification case.
-- PJ & RG
-- 2017-01-07: ported from the Eilmer3 example

-- We can set individual attributes of the global data object.
config.title = "Oblique detonation wave with Powers-Aslam gas model."
print(config.title)
config.dimensions = 2

nsp, nmodes, gm = setGasModel('powers-aslam-gas-model.lua')
print("GasModel has nsp= ", nsp, " nmodes= ", nmodes)
massf1 = {A=1.0, B=0.0}
initial = FlowState:new{p=28.7e3, T=300.0, massf=massf1}
inflow = FlowState:new{p=86.1e3, T=300.0, velx=964.302, massf=massf1}

-- Geometry
xmin = -0.25
xmax = 1.75
ymin = 0.0
ymax = 2.0

dofile("analytic.lua")
myWallFn = create_wall_function(0.0, xmax)

-- Set up two patches in the (x,y)-plane by first defining
-- the corner nodes, then the lines between those corners.
a = Vector3:new{x=xmin, y=0.0}
b = Vector3:new{x=0.0, y=0.0}
c = Vector3:new{x=myWallFn(1.0).x, y=myWallFn(1.0).y}
d = Vector3:new{x=xmin, y=ymax}
e = Vector3:new{x=0.0, y=ymax}
f = Vector3:new{x=xmax, y=ymax}
south0 = Line:new{p0=a, p1=b} -- upstream of wedge
south1 = LuaFnPath:new{luaFnName="myWallFn"} -- wedge surface
north0 = Line:new{p0=d, p1=e}; north1 = Line:new{p0=e, p1=f}
west0 = Line:new{p0=a, p1=d} -- inflow boundary
east0west1 = Line:new{p0=b, p1=e} -- vertical line, between patches
east1 = Line:new{p0=c, p1=f} -- outflow boundary
patch0 = makePatch{north=north0, east=east0west1, south=south0, west=west0}
patch1 = makePatch{north=north1, east=east1, south=south1, west=east0west1}
-- Mesh the patches, with particular discretisation.
factor = 2 -- for adjusting the grid resolution
nxcells = math.floor(40*factor)
nycells = math.floor(40*factor)
fraction0 = (0-xmin)/(xmax-xmin) -- fraction of domain upstream of wedge
nx0 = math.floor(fraction0*nxcells); nx1 = nxcells-nx0; ny = nycells
grid0 = StructuredGrid:new{psurface=patch0, niv=nx0+1, njv=ny+1}
grid1 = StructuredGrid:new{psurface=patch1, niv=nx1+1, njv=ny+1}
-- Define the flow-solution blocks and set boundary conditions.
-- We split the patches into roughly equal blocks so that
-- we make good use of our multicore machines.
blk0 = FBArray:new{grid=grid0, initialState=inflow, nib=1, njb=2,
                   bcList={west=InFlowBC_Supersonic:new{flowState=inflow},
                           north=OutFlowBC_Simple:new{}}}
blk1 = FBArray:new{grid=grid1, initialState=initial, nib=7, njb=2,
                   bcList={east=OutFlowBC_Simple:new{},
                           north=OutFlowBC_Simple:new{}}}
identifyBlockConnections()

-- Do a little more setting of global data.
config.reacting = true
config.reactions_file = "powers-aslam-gas-model.lua"
config.flux_calculator = "ausmdv"
config.max_time = 2.0e-2  -- seconds
config.max_step = 300000
config.dt_init = 1.0e-6
config.dt_plot = config.max_time / 40
config.dt_history = 10.0e-5

dofile("sketch-domain.lua")

File: examples/lorikeet/oblique-detonation-wave/odw.py
# odw.py
# Oblique detonation wave with a simple reacting gas.
# Modelled on Rowan's example for Eilmer3, way back in the year 2006.
# Peter J. 2022-12-29 Adapted from the Chicken example.
#
config.title = "Oblique detonation wave with Powers-Aslam gas model."
print(config.title)
#
init_gas_model("powers-aslam-gas-model.lua")
inflow = FlowState(p=86.1e3, T=300.0, velx=964.302, massf=[1.0, 0.0])
initial = FlowState(p=28.7e3, T=300.0, velx=0.0, massf=[1.0, 0.0])
config.reacting = True
#
# Geometry
from gdtk.geom.path import Line, FnPath
from gdtk.geom.surface import CoonsPatch
#
xmin = -0.25; xmax = 1.75
ymin = 0.0; ymax = 2.0
#
from oblique_detonation import *
from math import pi
od =  ObliqueDetonation(pi/4.0, 300.0, 3.0, 1.0)
wall = FnPath(od.create_wall_function(0.0, xmax))
#
# We are going to build two patches in the x,y plane.
a0 = Vector3(xmin, 0.0); a1 = Vector3(xmin, ymax)
b0 = Vector3(0.0, 0.0);  b1 = Vector3(0.0, ymax)
c0 = wall(1.0); c1 = Vector3(xmax, ymax)
#
patch0 = CoonsPatch(p00=a0, p10=b0, p11=b1, p01=a1)
patch1 = CoonsPatch(west=Line(b0,b1), east=Line(c0,c1),
                    north=Line(b1,c1), south=wall)
#
# Mesh the patches, with particular discretisation.
#
factor = 2
nn = factor*40
nnx = nn; nny = nn # overall numbers of cells
fraction0 = (0-xmin)/(xmax-xmin) # fraction of domain upstream of wedge
nnx0 = int(fraction0*nnx) # share for blk0
nnx1 = nnx - nnx0 # share for blk1
grd0 = StructuredGrid(psurf=patch0, niv=nnx0+1, njv=nny+1)
grd1 = StructuredGrid(psurf=patch1, niv=nnx1+1, njv=nny+1)
#
blk0 = FluidBlock(i=0, grid=grd0, initialState=inflow,
                  bcs={'iminus':InflowBC(inflow),'iplus':ExchangeBC()})
blk1 = FluidBlock(i=1, grid=grd1, initialState=initial,
                  bcs={'iminus':ExchangeBC(),'iplus':OutflowBC()})
#
config.flux_calc = FluxCalc.ausmdv_plus_hanel
config.max_time = 2.0e-2
config.max_step = 300000
config.plot_dt = config.max_time/40
